File: tests/human/x/pas/dataset_sort_take_limit.pas
program DatasetSortTakeLimit;

type
  TProduct = record
    name: string;
    price: integer;
  end;

var
  products: array[1..7] of TProduct = (
    (name:'Laptop'; price:1500),
    (name:'Smartphone'; price:900),
    (name:'Tablet'; price:600),
    (name:'Monitor'; price:300),
    (name:'Keyboard'; price:100),
    (name:'Mouse'; price:50),
    (name:'Headphones'; price:200)
  );
  i,j: integer;
  temp: TProduct;
begin
  { sort by price descending }
  for i := 1 to High(products)-1 do
    for j := i+1 to High(products) do
      if products[i].price < products[j].price then
      begin
        temp := products[i];
        products[i] := products[j];
        products[j] := temp;
      end;

  Writeln('--- Top products (excluding most expensive) ---');
  for i := 2 to 4 do
    Writeln(products[i].name, ' costs $', products[i].price);
end.

File: tests/human/x/fortran/dataset_sort_take_limit.f90
program dataset_sort_take_limit
  implicit none
  type :: Product
    character(len=20) :: name
    integer :: price
  end type Product
  type(Product) :: products(7)
  type(Product) :: sorted(7)
  integer :: i, j
  type(Product) :: temp

  products(1) = Product('Laptop',1500)
  products(2) = Product('Smartphone',900)
  products(3) = Product('Tablet',600)
  products(4) = Product('Monitor',300)
  products(5) = Product('Keyboard',100)
  products(6) = Product('Mouse',50)
  products(7) = Product('Headphones',200)

  sorted = products
  do i = 1, 6
    do j = i+1, 7
      if (sorted(j)%price > sorted(i)%price) then
        temp = sorted(i)
        sorted(i) = sorted(j)
        sorted(j) = temp
      end if
    end do
  end do

  print *, '--- Top products (excluding most expensive) ---'
  do i = 2, 4
    print *, trim(sorted(i)%name)//' costs $', sorted(i)%price
  end do
end program dataset_sort_take_limit
